File: src/main/scala/category/stack/_456_Solution.scala
package category.stack

import java.util

/**
  * 456. 132 Pattern
  *
  * [1,2,3,4]
  * [3, 1, 4, 2]
  * [-1, 3, 2, 0]
  * []
  * [1,2]
  * [6,12,3,4,6,20,11]
  * [6,12,3,4,6,11,20]
  */
object _456_Solution {
  def find132pattern(nums: Array[Int]): Boolean = {
    val N = nums.length
    if (N < 3) return false
    val minArray = new Array[Int](N)
    minArray(0) = nums(0)
    for (i <- 1 until N) minArray(i) = Math.min(minArray(i - 1), nums(i))
    val stack = new util.Stack[Int]
    for (j <- Range(N - 1, -1, -1)) {
      while (!stack.isEmpty && stack.peek() <= minArray(j)) stack.pop()
      if (!stack.isEmpty && stack.peek() < nums(j)) return true
      stack.push(nums(j))
    }
    false
  }
}

File: src/main/python/medium/_450_500/_456_Solution.py
# 456. 132 Pattern
from typing import List

'''
[1,2,3,4]
[1,3,-1,2]
[1,3,-1,4,3]
[1]
[2]
[1,3]
[3,1]
[1,3,2]
[1,1,1,3,2]
[1,-1,-1,3,-1,-1,2]
'''


class _456_Solution:
    # reversed
    def find132pattern(self, nums: List[int]) -> bool:
        if not nums: return False
        st, s3 = [], -0xffffffff
        for i in reversed(range(len(nums))):
            num = nums[i]
            if num < s3:
                return True
            else:
                while len(st) > 0 and num > st[-1]: s3 = st.pop()
            st.append(num)
        return False

    # order
    # def find132pattern(self, nums: List[int]) -> bool:
    #     if not nums: return False
    #     st, s1 = [], 0xffffffff
    #     for num in nums:
    #         if num > s1 and num < st[-1]:
    #             return True
    #         else:
    #             while len(st) > 0 and num >= st[-1]: s1 = min(s1, st.pop())
    #         st.append(num)
    #     return False


if __name__ == '__main__':
    obj = _456_Solution()
    assert obj.find132pattern([1, -1, -1, 3, -1, -1, 2]) == True
    assert obj.find132pattern([-2, 1, 1]) == False
    assert obj.find132pattern([8, 10, 4, 6, 5]) == True
    assert obj.find132pattern([1, 1, 1, 1, 1, 1, 1, 1, 1]) == False
    assert obj.find132pattern([1, 3, -1, 2]) == True
    assert obj.find132pattern([1, 3, 2]) == True
